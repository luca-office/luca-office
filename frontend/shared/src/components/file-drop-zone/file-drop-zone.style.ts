import {css} from "@emotion/react"
import * as React from "react"
import {
  backgroundColorLight,
  borderRadiusLarge,
  dashedBorderColor,
  errorColor,
  Flex,
  fontColor,
  FontWeight,
  primaryColor,
  spacingMedium,
  spacingSmall,
  spacingTiny
} from "../../styles"
import {DragState} from "./file-drop-zone"

const getColor = (props: DragState) => {
  if (props.isDragAccepted) {
    return primaryColor
  }
  if (props.isDragRejected) {
    return errorColor
  }
  if (props.isDragActive) {
    return primaryColor
  }
  return dashedBorderColor
}

const Size = {
  hint: 372
}

export const fileDropZoneHighlightStyle: React.CSSProperties = {
  fontWeight: "bold"
}

export const fileDropZoneStyle = {
  dragWrapper: (props: DragState, disabled: boolean) =>
    css(Flex.column, {
      flex: "1",
      alignItems: "center",
      padding: spacingMedium,
      cursor: disabled ? "not-allowed" : "pointer",
      borderWidth: "2px",
      borderRadius: borderRadiusLarge,
      borderColor: getColor(props),
      borderStyle: "dashed",
      backgroundColor: backgroundColorLight,
      color: "#bdbdbd",
      outline: "none",
      transition: "border 0.24s ease-in-out"
    }),
  selectFileButton: (isDisabled: boolean) =>
    css({
      color: isDisabled ? fontColor : primaryColor,
      fontWeight: FontWeight.Bold,
      marginTop: spacingSmall
    }),
  errorText: css({
    color: errorColor,
    marginTop: spacingTiny
  }),
  hint: css({
    maxWidth: Size.hint
  })
}
