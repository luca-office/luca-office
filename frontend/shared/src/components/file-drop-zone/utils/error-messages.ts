import {LucaI18nLangKey} from "../../../translations"

export enum FileUploadError {
  FileTooLarge,
  InvalidFileType,
  TooManyFiles
}

export const getLanguageKeyForFileUploadError = (error: FileUploadError): LucaI18nLangKey => {
  switch (error) {
    case FileUploadError.FileTooLarge:
      return "files_and_directories__upload_modal_file_too_large"
    case FileUploadError.InvalidFileType:
      return "files_and_directories__upload_modal_file_type_invalid"
    case FileUploadError.TooManyFiles:
      return "files_and_directories__upload_modal_no_single_file"
  }
}

export type FileErrorMessage = "file-too-large" | "file-too-small" | "too-many-files" | "file-invalid-type"

export const mapDropzoneRejectionToFileUploadError = (errorMessage: FileErrorMessage) => {
  switch (errorMessage) {
    case "file-invalid-type":
      return FileUploadError.InvalidFileType
    case "file-too-large":
      return FileUploadError.FileTooLarge
    case "too-many-files":
      return FileUploadError.TooManyFiles
    default:
      return FileUploadError.InvalidFileType
  }
}
