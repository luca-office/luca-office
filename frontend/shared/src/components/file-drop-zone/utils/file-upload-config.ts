export const maximumFileSizeInMegaByte = 100
export const maximumFileSizeInBytes = maximumFileSizeInMegaByte * Math.pow(10, 6)
