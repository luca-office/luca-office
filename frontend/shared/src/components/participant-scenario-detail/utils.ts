import {FreetextAnswerFragment} from "../../graphql/generated/FreetextAnswerFragment"
import {QuestionType} from "../../graphql/generated/globalTypes"
import {Questionnaire, RuntimeSurveyResults} from "../../models"
import {ParticipantQuestionAnswer, QuestionnaireParticipantResults} from "../questionnaire"

export const runtimeSurveyResultsToQuestionnaireParticipantResults = (
  invitationId: UUID,
  questionnaire: Questionnaire,
  runtimeSurveyResults: RuntimeSurveyResults,
  freeTextAnswers: FreetextAnswerFragment[]
): QuestionnaireParticipantResults => {
  const answers = runtimeSurveyResults.questionResults.reduce<Record<UUID, ParticipantQuestionAnswer>>(
    (answersAccumulator, result) => {
      const question = questionnaire.questions.find(question => question.id === result.questionId)
      const participantResult = result.participantResults.find(results => results.invitationId === invitationId)
      const freeTextAnswer = freeTextAnswers.find(
        answer => answer.questionId === result.questionId && answer.surveyInvitationId === invitationId
      )?.text

      switch (question?.questionType) {
        case QuestionType.SingleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.SingleChoice,
              questionId: question.id,
              selectedAnswerId: participantResult?.selectedAnswerIds[0] ?? "",
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.MultipleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.MultipleChoice,
              questionId: question.id,
              selectedAnswerIds: participantResult?.selectedAnswerIds ?? [],
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.FreeText: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.FreeText,
              questionId: question.id,
              freeTextAnswer: freeTextAnswer ?? ""
            }
          }
        }

        default:
          return answersAccumulator
      }
    },
    {}
  )

  return {questionnaireId: runtimeSurveyResults.questionnaireId, answers}
}
