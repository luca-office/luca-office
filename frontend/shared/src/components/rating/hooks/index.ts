export * from "./use-coding-criteria-by-items-list"
export * from "./use-free-text-answers-for-participants"
export * from "./use-freetext-question-ratings-by-ratings-list"
export * from "./use-questionnaire-rating-score"
export * from "./use-rating"
export * from "./use-ratings"
export * from "./use-scenario-coding-item-ratings-by-ratings-list"
export * from "./use-scenario-coding-items-by-ratings-list"
export * from "./use-scenario-rating-score"
export * from "./use-selected-answers-for-participant-by-question-list"
export * from "./use-selected-answers-for-participant-by-survey-invitation-list"
export * from "./use-selected-answers-for-participant-by-survey-invitations-list"
