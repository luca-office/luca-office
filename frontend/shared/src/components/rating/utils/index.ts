export * from "./coding-criterion"
export * from "./coding-dimension"
export * from "./coding-item"
export * from "./question"
export * from "./score"
export * from "./scoring-type"
