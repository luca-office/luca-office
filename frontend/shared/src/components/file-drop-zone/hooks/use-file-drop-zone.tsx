import {useState} from "react"
import {Accept, DropzoneState, FileRejection, useDropzone} from "react-dropzone"
import {Option} from "../../../utils"
import {FileErrorMessage, FileUploadError, mapDropzoneRejectionToFileUploadError} from "../utils/error-messages"
import {maximumFileSizeInBytes} from "../utils/file-upload-config"

export interface UseFileDropZoneHook {
  readonly dropzone: DropzoneState
  readonly fileUploadError: Option<FileUploadError>
}

export interface UseFileDropParams {
  readonly accept?: Accept
  readonly onFilesAccepted: (files: File[]) => void
  readonly multiple?: boolean
}

export const useFileDropzone = (params: UseFileDropParams): UseFileDropZoneHook => {
  const handleFileRejection = (fileRejections: FileRejection[]) => {
    if (fileRejections.length > 0) {
      const fileUploadErrors = fileRejections[0].errors

      if (fileUploadErrors.length > 0) {
        setFileUploadError(
          Option.of(mapDropzoneRejectionToFileUploadError(fileUploadErrors[0].code as FileErrorMessage))
        )
      }
    }
  }

  const handleDropAccept = (files: File[]) => {
    setFileUploadError(Option.none())
    params.onFilesAccepted(files)
  }

  const dropzone = useDropzone({
    accept: params.accept,
    maxSize: maximumFileSizeInBytes,
    onDropAccepted: handleDropAccept,
    onDropRejected: handleFileRejection,
    multiple: params.multiple ?? true,
    noKeyboard: true
  })

  const [fileUploadError, setFileUploadError] = useState<Option<FileUploadError>>(Option.none())

  return {
    dropzone,
    fileUploadError
  }
}
