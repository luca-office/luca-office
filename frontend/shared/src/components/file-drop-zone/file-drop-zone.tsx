import * as React from "react"
import {Accept} from "react-dropzone"
import Highlighter from "react-highlight-words"
import {CustomStyle} from "../../styles"
import {useLucaTranslation} from "../../translations"
import {Text} from "../typography/typography"
import {fileDropZoneHighlightStyle as highlightStyle, fileDropZoneStyle as styles} from "./file-drop-zone.style"
import {useFileDropzone} from "./hooks/use-file-drop-zone"
import {getLanguageKeyForFileUploadError} from "./utils/error-messages"
import {maximumFileSizeInMegaByte} from "./utils/file-upload-config"

export interface DragState {
  readonly isDragActive: boolean
  readonly isDragAccepted: boolean
  readonly isDragRejected: boolean
}

export interface FileDropZoneProps extends CustomStyle {
  readonly accept?: Accept
  readonly onFilesAccepted: (files: File[]) => void
  readonly multiple?: boolean
  readonly disabled?: boolean
  readonly description?: React.ReactNode
  readonly hideSelectButton?: boolean
}

export const FileDropzone: React.FC<FileDropZoneProps> = ({
  accept,
  onFilesAccepted,
  multiple = true,
  disabled,
  description,
  hideSelectButton = false,
  customStyles
}) => {
  const {dropzone, fileUploadError} = useFileDropzone({accept, onFilesAccepted, multiple})
  const {t} = useLucaTranslation()
  const {isDragAccept, isDragReject, isDragActive, getInputProps, getRootProps} = dropzone

  return (
    <>
      <div
        className="file-dropzone-wrapper"
        css={[
          styles.dragWrapper({isDragAccepted: isDragAccept, isDragActive, isDragRejected: isDragReject}, !!disabled),
          customStyles
        ]}
        {...getRootProps()}>
        <input {...getInputProps({multiple, disabled})} />
        <Text customStyles={styles.hint}>
          {description ?? (
            <Highlighter
              highlightTag={"span"}
              highlightStyle={highlightStyle}
              searchWords={[`${maximumFileSizeInMegaByte} MB`]}
              textToHighlight={t(
                multiple
                  ? "files_and_directories__upload_modal_file_size_hint_plural"
                  : "files_and_directories__upload_modal_file_size_hint_singular",
                {
                  fileSizeInMB: maximumFileSizeInMegaByte
                }
              )}
            />
          )}
        </Text>
        {!hideSelectButton && (
          <Text customStyles={styles.selectFileButton(!!disabled)}>
            {t(
              disabled
                ? "files_and_directories__upload_modal_choose_file_disabled"
                : "files_and_directories__upload_modal_choose_file"
            )}
          </Text>
        )}
      </div>
      {fileUploadError
        .map(error => <Text customStyles={styles.errorText}>{t(getLanguageKeyForFileUploadError(error))}</Text>)
        .orNull()}
    </>
  )
}
