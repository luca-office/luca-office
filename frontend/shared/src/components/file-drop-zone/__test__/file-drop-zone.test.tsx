import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Option} from "../../../utils"
import {Text} from "../../typography/typography"
import {FileDropzone, FileDropZoneProps} from "../file-drop-zone"
import * as fileDropZoneHook from "../hooks/use-file-drop-zone"
import {UseFileDropZoneHook} from "../hooks/use-file-drop-zone"
import {FileUploadError} from "../utils/error-messages"

const defaultProps: FileDropZoneProps = {
  onFilesAccepted: jest.fn(),
  accept: {
    "image/*": [".jpg", ".svg", ".png", ".jpeg", ".gif"]
  }
}

const hookValuesDefault: UseFileDropZoneHook = {
  dropzone: {
    open: jest.fn(),
    isFocused: false,
    isDragReject: false,
    acceptedFiles: [],
    fileRejections: [],
    getInputProps: jest.fn(),
    getRootProps: jest.fn(),
    inputRef: {current: null},
    isDragAccept: true,
    isDragActive: false,
    isFileDialogActive: false,
    rootRef: {current: null}
  },
  fileUploadError: Option.none()
}

const stateSpy = jest.spyOn(fileDropZoneHook, "useFileDropzone")

const getComponent = (props?: Partial<FileDropZoneProps>) => <FileDropzone {...{...defaultProps, ...props}} />

describe("file-drop-zone", () => {
  it("renders correctly", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("renders correctly (hide select button)", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent({hideSelectButton: true})
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = getComponent()
    const tree = shallow(component)
    expect(tree.find(".file-dropzone-wrapper")).toHaveLength(1)
    expect(tree.find(Text)).toHaveLength(2)
  })
  it("has correct structure with error", () => {
    stateSpy.mockReturnValue({
      ...hookValuesDefault,
      fileUploadError: Option.of<FileUploadError>(FileUploadError.FileTooLarge)
    })
    const component = getComponent({})
    const tree = shallow(component)
    expect(tree.find(".file-dropzone-wrapper")).toHaveLength(1)
    expect(tree.find(Text)).toHaveLength(3)
  })
})
