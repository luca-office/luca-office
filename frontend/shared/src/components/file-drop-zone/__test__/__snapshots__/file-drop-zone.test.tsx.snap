// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file-drop-zone renders correctly (hide select button) 1`] = `
<div
  className="file-dropzone-wrapper css-1lbthkj-FileDropzone"
>
  <input />
  <div
    className="css-vq2l7y-Text"
  >
    <span>
      <span
        className=""
      >
        files_and_directories__upload_modal_file_size_hint_plural
      </span>
    </span>
  </div>
</div>
`;

exports[`file-drop-zone renders correctly 1`] = `
<div
  className="file-dropzone-wrapper css-1lbthkj-FileDropzone"
>
  <input />
  <div
    className="css-vq2l7y-Text"
  >
    <span>
      <span
        className=""
      >
        files_and_directories__upload_modal_file_size_hint_plural
      </span>
    </span>
  </div>
  <div
    className="css-1oed4f3-Text"
  >
    files_and_directories__upload_modal_choose_file
  </div>
</div>
`;
